interfaces/apparmor/backend_test.go

```diff
@@ -0,0 +1,358 @@
+			fmt.Sprintf("--replace --write-cache -O no-expr-simplify --cache-loc=/var/cache/apparmor %s", profile),
+		})
+		s.removeSnap(c, snapInfo)
+	}
+}
+
+func (s *backendSuite) TestUpdatingSnapToOneWithMoreApps(c *C) {
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, sambaYamlV1)
+		s.cmds["apparmor_parser"].ForgetCalls()
+		snapInfo = s.updateSnap(c, snapInfo, developerMode, sambaYamlV1WithNmbd)
+		profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.nmbd")
+		// file called "snap.sambda.nmbd" was created
+		_, err := os.Stat(profile)
+		c.Check(err, IsNil)
+		// apparmor_parser was used to load the new profile
+		c.Check(s.cmds["apparmor_parser"].Calls(), DeepEquals, []string{
+			fmt.Sprintf("--replace --write-cache -O no-expr-simplify --cache-loc=/var/cache/apparmor %s", profile),
+		})
+		s.removeSnap(c, snapInfo)
+	}
+}
+
+func (s *backendSuite) TestUpdatingSnapToOneWithFewerApps(c *C) {
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, sambaYamlV1WithNmbd)
+		s.cmds["apparmor_parser"].ForgetCalls()
+		snapInfo = s.updateSnap(c, snapInfo, developerMode, sambaYamlV1)
+		profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.nmbd")
+		// file called "snap.sambda.nmbd" was removed
+		_, err := os.Stat(profile)
+		c.Check(os.IsNotExist(err), Equals, true)
+		// apparmor_parser was used to remove the unused profile
+		c.Check(s.cmds["apparmor_parser"].Calls(), DeepEquals, []string{"--remove snap.samba.nmbd"})
+		s.removeSnap(c, snapInfo)
+	}
+}
+
+func (s *backendSuite) TestRealDefaultTemplateIsNormallyUsed(c *C) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(sambaYamlV1))
+	c.Assert(err, IsNil)
+	// NOTE: we don't call apparmor.MockTemplate()
+	err = s.backend.Configure(snapInfo, false, s.repo)
+	c.Assert(err, IsNil)
+	profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.smbd")
+	data, err := ioutil.ReadFile(profile)
+	c.Assert(err, IsNil)
+	for _, line := range []string{
+		// NOTE: a few randomly picked lines from the real profile.  Comments
+		// and empty lines are avoided as those can be discarded in the future.
+		"#include <tunables/global>\n",
+		"/tmp/   r,\n",
+		"/sys/class/ r,\n",
+	} {
+		c.Assert(string(data), testutil.Contains, line)
+	}
+}
+
+func (s *backendSuite) TestCustomTemplateUsedOnRequest(c *C) {
+	s.backend.CustomTemplate = `
+# Description: Custom template for testing
+###VAR###
+
+###PROFILEATTACH### (attach_disconnected) {
+	FOO
+}
+`
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(sambaYamlV1))
+	c.Assert(err, IsNil)
+	err = s.backend.Configure(snapInfo, false, s.repo)
+	c.Assert(err, IsNil)
+	profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.smbd")
+	data, err := ioutil.ReadFile(profile)
+	c.Assert(err, IsNil)
+	// Our custom template was used
+	c.Assert(string(data), testutil.Contains, "FOO")
+	// Custom profile can rely on legacy variables
```

Comment:
Because they originate from 15.04 and we'll probably simplify things for 16.04 so that we use just the concepts of snap name, app name and security tag. This is still open for discussion, I just included it here. As a second reason "legacy variables" are always used for old-security since it has to be compatible with earlier releases.